urlb7d18764c0_steveliles_health_tgz-pJ8-com_sjl_health_SampleConfigurationJ8/src/com/sjl/health/internal/DynamicProxyingHealthService.java

```diff
@@ -0,0 +1,158 @@
+		instrumentedClasses = new WeakHashMap<Class<?>, Map<Method, Handler>>();
+	}
+	
+	@Override
+	@SuppressWarnings("unchecked")
+	public <T> T monitor(final T aComponent, Configuration aConfig) {
+		aConfig.init(issueTrackerFactory, clock);
+System.out.println(aConfig);		
+		final Map<Method, Handler> _handlers = getHandlers(aComponent.getClass());
+		final Health _health = healthFactory.newHealth(aConfig);
+		
+		T _instrumented = (T) Proxy.newProxyInstance(
+			aComponent.getClass().getClassLoader(), 
+			aComponent.getClass().getInterfaces(), 
+            new InvocationHandler() {
+            @Override
+            public Object invoke(Object aProxy, Method aMethod, Object[] aArgs) 
+            throws Throwable {            	
+            	try {
+            		Handler _h = _handlers.get(aMethod);            		
+            		if (_h != null)
+            			return _h.handle(_health, aComponent, aArgs);
+            		return aMethod.invoke(aComponent, aArgs);
+            	} catch (InvocationTargetException anExc) {          		
+            		throw anExc.getCause();
+            	}
+            }
+        });
+		
+		addComponent(aComponent, _health);
+		
+		return _instrumented;
+	}
+
+	@Override
+	public HealthInfo get(Object aMaybeMonitored) {
+		synchronized (lock) {
+			return monitoredComponents.get(aMaybeMonitored);
+		}
+	}
+	
+	public void reset(Object aMaybeMonitored) {
+		synchronized(lock) {
+			Health _h = monitoredComponents.get(aMaybeMonitored);
+			if (_h != null)
+				_h.reset();
+		}
+	}
+	
+	private void addComponent(Object aComponent, Health aHealth) {
+		synchronized(lock) {
+			monitoredComponents.put(aComponent, aHealth);
+		}
+	}
+	
+	private Map<Method, Handler> getHandlers(Class<?> aComponentClass) {
+		synchronized(lock) {
+			Map<Method, Handler> _handlers = instrumentedClasses.get(aComponentClass);
+			if (_handlers == null) {
+				_handlers = initHandlers(aComponentClass);
+				instrumentedClasses.put(aComponentClass, _handlers);
+			}
+			return _handlers;
+		}
+	}
+	
+	private Map<Method, Handler> initHandlers(Class<?> aComponentClass) {
+		Map<Method, Handler> _result = new HashMap<Method, Handler>();
+		// only instrument methods of implemented interfaces
+		for (Class<?> _cls : aComponentClass.getInterfaces()) {
+			for (Method _m : _cls.getMethods()) {
+				if (_m.isAnnotationPresent(NotInstrumented.class)) {
+					_result.put(_m, newUninstrumentedHandler(_m));
+				} else if (_m.isAnnotationPresent(InstrumentedResult.class)) {
+					_result.put(_m, newInstrumentedResultHandler(_m));
+				} else {
+					_result.put(_m, newInstrumentedHandler(_m));
+				}
+			}
+		}
+		return _result;
+	}
+	
+	private Handler newUninstrumentedHandler(final Method aMethod) {
+		return new Handler() {
+            @Override
+            public Object handle(OperationMonitor aMonitor, Object aProxied, Object[] anArgs) 
+            throws Throwable {
+                return aMethod.invoke(aProxied, anArgs);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)